lib/os/mem_blocks.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/__assert.h>
+#include <sys/check.h>
+#include <sys/mem_blocks.h>
+#include <sys/util.h>
+
+static void *alloc_one(sys_mem_blocks_t *mem_block)
+{
+	size_t offset;
+	int r;
+	uintptr_t uptr;
+	void *ret = NULL;
+
+	/* Find an unallocated block */
+	r = sys_bitarray_alloc(mem_block->bitmap, 1, &offset);
+	if (r != 0) {
+		goto out;
+	}
+
+	/* Calculate the start address of the newly allocated block */
+	uptr = POINTER_TO_UINT(mem_block->buffer) + (offset * mem_block->block_size);
+
+	ret = UINT_TO_POINTER(uptr);
+
+out:
+	return ret;
+}
+
+static int free_one(sys_mem_blocks_t *mem_block, void *ptr)
+{
+	size_t offset;
+	uintptr_t uptr = POINTER_TO_UINT(ptr);
+	int ret = 0;
+
+	/* Make sure ptr is within the mem_block buffer */
+	if (uptr < POINTER_TO_UINT(mem_block->buffer)) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	offset = (uptr - POINTER_TO_UINT(mem_block->buffer)) / mem_block->block_size;
+	if (offset >= mem_block->num_blocks) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	ret = sys_bitarray_free(mem_block->bitmap, 1, offset);
+
+out:
+	return ret;
+}
+
+int sys_mem_blocks_alloc(sys_mem_blocks_t *mem_block, size_t count,
+			 uintptr_t *out_blocks)
+{
+	int ret = 0;
+	int i;
+
+	if ((mem_block == NULL) || (out_blocks == NULL)) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	CHECKIF((mem_block->bitmap == NULL) || (mem_block->buffer == NULL)) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	if (count == 0) {
+		/* Nothing to allocate */
+		goto out;
+	}
+
+	if (count > mem_block->num_blocks) {
+		/* Definitely not enough blocks to be allocated */
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	for (i = 0; i < count; i++) {
```

Comment:
This loop with the following free could I suppose potentially lead to some allocation races where two cores contend on allocating some number of blocks, and neither succeeds to allocate enough blocks despite enough being there for either of the calls... Nothing necessarily bad comes from it, but maybe not the most friendly outcome if that's a common use case